src/librustc_typeck/check_unused.rs

```diff
@@ -72,4 +74,16 @@ pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>) {
 
     let mut visitor = CheckVisitor { tcx, used_trait_imports };
     tcx.hir.krate().visit_all_item_likes(&mut visitor);
+
+    for &(id, span, cnum) in &tcx.maybe_unused_extern_crates {
+        let metadata = tcx.sess.cstore.crate_data_as_rc_any(cnum);
+        let metadata = metadata.downcast_ref::<CrateMetadata>()
+            .expect("CrateStore crate data is not a CrateMetadata");
+        if !metadata.is_compiler_builtins(&tcx.dep_graph)
+            && !metadata.is_panic_runtime(&tcx.dep_graph) {
```

Comment:
Two problems here:

First, we should be checked for global allocator crates as well. Otherwise, if you have the following setup, we will warn incorrectly:

- Crate A defines a `#[global_allocator] static ALLOCATOR`.
- Crate B uses Crate A just for this side-effect.

I think this is not the recommended pattern, but it still seems *legal* to me, and I think we should capture it rather than giving a false warning.

Second, we should not be downcasting to `CrateMetadata` -- more generally, `librustc_typeck` should not directly depend on `librustc_metadata`. Rather, these calls should be going through queries. For example, the line that says `!metadata.is_panic_runtime(&tcx.dep_graph)` should be as follows:

```
!tcx.is_panic_runtime(cnum.as_def_id())
```

The problem is that, at the moment, only the `is_panic_runtime` query exists, so you'll have to add the others. This is basically a matter of copy-and-pasting the `is_panic_runtime` code you find here:

- [add it to the macro in `ty/maps.rs`](https://github.com/rust-lang/rust/blob/5431f423aac345e023815d3f10b8cb840711d756/src/librustc/ty/maps.rs#L997),
- add an [implementation in `cstore_impl.rs`](https://github.com/rust-lang/rust/blob/5431f423aac345e023815d3f10b8cb840711d756/src/librustc_metadata/cstore_impl.rs#L139).

You'll want to do that for `is_compiler_builtins` and `has_global_allocator`.

Doing things through queries is the recommend style in general, but it also means that this code will play properly with incremental compilation and so forth.

(Now that I think about it, there might *still* be a bug here, in that I'm not sure if the `has_global_allocator` flag means that it has an *immediate* global allocator, or if it covers the transitive case too, but I think it's good enough for now.)